### ISSUE TITLE
Incorrectly reported file status for symlinked files on Windows

### ISSUE THREAD

**Issue by edmorley** [sublimehq/sublime_merge#292]
**Version info**

 - OS: Windows 10 Pro build 17763
 - Build: 1094

**Description**

Sublime Merge reports incorrect "File mode changed: Symbolic Link -> Normal File" working directory changes when other tools (such as Source Tree or command line `git` from Git for Windows) report no changes. This is presumably since Sublime Merge isn't handling `core.symlinks` correctly.

**Steps to reproduce**

1. Ensure using a recent Windows 10 build (that includes [this feature](https://blogs.windows.com/buildingapps/2016/12/02/symlinks-windows-10/))
2. Enable Window 10's developer mode ([guide](https://docs.microsoft.com/en-us/windows/uwp/get-started/enable-your-device-for-development))
3. Install Git for Windows from https://git-scm.com/download/win - making sure to accept the default Git config suggestions, which include setting `core.symlinks = true`
4. Using Git for Windows, clone https://github.com/neutrinojs/neutrino (since it contains symlinks under `docs/packages/`)
5. Open that repository in Sublime Merge, and check the reported working directory status

**Expected behavior**

Sublime Merge reports no changes in the working directory.

**Screenshots**

![sublime-merge-symlinks](https://user-images.githubusercontent.com/501702/49459304-abb36680-f7e6-11e8-8118-01c7bf877d4e.jpg)

**Comment by jskinner** [comment]
It looks like Sublime Merge is reading your repository with `core.symlinks=false`. You can verify this using "Help/Debug Information", which will show all used config values.

Sublime Merge uses its bundled version of Git by default, which has core.symlinks set to false. If you have a git repo cloned using the system version of git, you may want to configure Sublime Merge to use the system version of git in the preferences menu, in which case it'll start using the system configuration values.

**Comment by edmorley** [comment]
Hi! Thank you for the reply.

Looking at the debug info output, I see:

```
=== Git Config Information ===
...
core.symlinks=true

=== Our Config Information ===
Git Config Path Information
Using config path: /C/ProgramData/Git/config
Using config path: /C/Program Files/Sublime Merge/Git/mingw64/etc/gitconfig
Using config path: /C/Users/Ed/.config/git/config
Using config path: /C/Users/Ed/.gitconfig
Using config path: /C/Users/Ed/src/neutrino/.git/config

Our config output is identical to the Git config output
```

**Comment by stinos** [comment]
> Sublime Merge uses its bundled version of Git by default, which has core.symlinks set to false

The bundled version picks up my global .gitconfig as well though:

```
=== Git Version Information ===
Using Git: bundled (custom)

=== Git Config Information ===
<Empty Config Output>

=== Our Config Information ===
Git Config Path Information
...
Using config path: /C/Users/<user>/.gitconfig
...
Our config output is different to the Git config output
...
core.symlinks=true
...
```

**Comment by stinos** [comment]
Anyone knows whether there are plans with this? This is a bit annoying, not only because it provides information which isn't correct and clutters the UI (one particular [repository](https://github.com/micropython/micropython) I'm working with has about 30 symlinks so every actual change requires hunting it down a forest of unrelated things), but also because it disables convenient functionality like `Edit commit Contents`.

**Comment by jimon** [comment]
Yeah I'm facing similar issue, though some of files might not be symlinked:

![image](https://user-images.githubusercontent.com/1333661/68589561-77333a80-048c-11ea-974f-d41d0c456a74.png)

`git status` doesn't show any of them.

**Comment by broadwaylamb** [comment]
I'm experiencing this issue as well with the 1119 build.

**Comment by forrest79** [comment]
The same here on build 2020. SmartGit or `git` in cli works well. In debug information is `core.symlinks=true` for system and also for bundled GIT binaries. Unfortunately, I can't use Sublime Merge with this bug :-(

**Comment by dpjohnst** [comment]
Hi @forrest79 (and those still experiencing this issue).

Are you able to reproduce this issue with the steps provided in the initial bug report?
Does the issue persist when changing the Git binary to system (via the preferences)?

In the debug information, can you see: Our config output is identical to the Git config output (under the Our Config Information section)?

Thanks,
\- Dylan

**Comment by forrest79** [comment]
When I try to do clone in SublimeMerge, no symlinks are created in the system. When I do clone in a command line or in SmartGit - symlinks are created as expected.

With bundled GIT, I can see "Our config output is identical to the Git config output", with system GIT - "Our config output is different to the Git config output". But the behavior is the same for both - symlinks are shown as modified.

I have Windows 10 Home edition, not in developer mode, just with enabled symlinks via Polsedit app (https://www.southsoftware.com/polsedit.html). And I'm using a portable version of SublimeMerge.

If you need some additional testing, please let me know.

**Comment by dpjohnst** [comment]
Hi @forrest79,

Most likely what is happening is that Sublime Merge is not receiving the correct permissions to create symlinks.

Could perform the following steps on the repository with the issue:

1. Open the terminal / command line to the repository
2. Run `git config core.symlinks` and share the output

**Comment by dpjohnst** [comment]
@stinos @broadwaylamb @jimon Could you confirm whether this issue is still affecting you on the latest build (2020 for stable, 2022 for dev)?

**Comment by jimon** [comment]
@dpjohnst yes, here is a screenshot from 2022 build, 4 files are:

![image](https://user-images.githubusercontent.com/1333661/84627825-e1f79900-aee7-11ea-9470-63d3221c1fa3.png)

- All 4 items are symlinks on windows.
- `git config core.symlinks` is set to `true`.
- Running `git diff` doesn't show anything like that.

If you want, send me a build with more debug output and I can check it for you :)

**Comment by stinos** [comment]
@dpjohnst I'm glad to report I do not currently have the issue (tested with repository mentioned by OP). No idea however when it got resolved exactly; I'm on build 2022.

Slightly related: the issue it is still present in SublimeText though: it shows all links with the default blue dot meaning 'modified'.

**Comment by forrest79** [comment]
> Hi @forrest79,
> 
> Most likely what is happening is that Sublime Merge is not receiving the correct permissions to create symlinks.
> 
> Could perform the following steps on the repository with the issue:
> 
> 1. Open the terminal / command line to the repository
> 2. Run `git config core.symlinks` and share the output

Hi, this is from a command line (without admin rights), my system GIT binaries:

```
git config core.symlinks
true
```

**Comment by dpjohnst** [comment]
Hi all,

For those seeing this issue, could you run the following via bash and share the output.
`stat PATH_TO_SYMLINK` where PATH_TO_SYMLINK is the path to a file expected to be a symlink.

Thanks,
\- Dylan

**Comment by forrest79** [comment]
Hi,

`stat P:\POMO\Web\csfd\vendor\pmgdev\caching`

Output:

```
  File: P:\POMO\Web\csfd\vendor\pmgdev\caching -> ../../packages/internals/Caching
  Size: 32              Blocks: 0          IO Block: 65536  symbolic link
Device: eaf1dea4h/3941719716d   Inode: 2251799814127377  Links: 1
Access: (0777/lrwxrwxrwx)  Uid: (197609/forrest79)   Gid: (197609/forrest79)
Access: 2020-04-08 14:52:04.063792300 +0200
Modify: 2020-04-08 14:52:04.063792300 +0200
Change: 2020-04-08 14:52:04.063792300 +0200
 Birth: 2020-04-08 14:52:04.063792300 +0200
```

**Comment by bempelise** [comment]
I tried to create a minimal repo that demonstrates the issue. I created one in gihub but upon cloning in windows, there was no issue with pending changes in Sublime Merge. So I ported the repo in Gitlab and I managed to replicate.

The repo is here:
`https://gitlab.com/bempelise/symlink`

For reference the repo that works fine can be found here:
`https://github.com/bempelise/symlink`

**Comment by dpjohnst** [comment]
Hi @bempelise,

I still haven't had any luck replicating this issue via the repo provided unfortunately.
I have added some extra debugging info to the upcoming development build to investigate this further.

Thank you for your patience while this is investigated.

Kind regards,
\- Dylan

**Comment by dpjohnst** [comment]
Hi all,

Quick update from our end: we've identified the issue and have a fix incoming - stay tuned.

Thanks to everyone who have reported this issue and shared details to debug this.

Kind regards,
\- Dylan

**Comment by kodewdle** [comment]
Directory Junctions on Build 2025 are now always showing, the files in them stage correctly but the main link displays as "Untracked Files" which "delete" and "stage" do nothing for.

**Comment by dpjohnst** [comment]
Hi all,

I'm going to go ahead and close this issue as it's been fixed, but for those affected by the directory junction issue, please feel free to subscribe to #830 to follow along.

Thanks,
\- Dylan

**Comment by bempelise** [comment]
Unfortunately I can still see this issue on one of my machines (it has been successfully resolved in others for the same repo though...)

Is there anything that I may do to fix it?

**Comment by blaenk** [comment]
Old post but including here cause it's what showed up in the Google results.

If anyone else runs into this on Windows with Developer Mode enabled (so as to not require admin for cloning the repo with symlinks): I checked debug info and noticed this:

```
=== App Version Information ===
Build: 2068

…

Our config output is identical to the Git config output

=== Git Attributes Information ===
git check_attr --all output
api/src/main/resources/graphql

=== Our Modified Files Newline Normalisation and EOL Information ===
api/src/main/resources/graphql: unspecified - native

=== Our Modified Files Flag Information ===
Ignoring symlinks: 0
api/src/main/resources/graphql: Working - Symbolic Link : Index - Symbolic Link
```

Despite the fact that when I ran that `git` command—even the one bundled with Sublime Merge—and did not observe this:

```
C:\Users\jorge\Code\app>"C:\Program Files\Sublime Merge\Git\cmd\git.exe" check-attr --all output
C:\Users\jorge\Code\app>
```

I also checked the bash `stat` output and also tried to look at the Windows file permissions, and there seemed to be no difference.

The weird thing is that I had the same repo albeit cloned with admin that did _not_ have this issue. The problem was that I was trying to clone as user, since I shouldn't have to require admin.

I tried a few things to make this "unstick".

In the end it seemed to get fixed by simply deleting the symlink, noting Sublime Merge detects the deletion, then within Sublime Merge, discard that deletion.
